source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -43,6 +83,132 @@ void Cluster::startPreInit() {
+        auto cluster_name = it->first;
+        ENVOY_LOG(debug, "cluster='{}' removing from cluster_map & cluster manager", cluster_name);
+        cluster_map_.erase(it++);
+        cm_.removeCluster(cluster_name);
+      } else {
+        ++it;
+      }
+    }
+  }
+  idle_timer_->enableTimer(sub_cluster_ttl_);
+}
+
+std::pair<bool, absl::optional<envoy::config::cluster::v3::Cluster>>
+Cluster::createSubClusterConfig(const std::string& cluster_name, const std::string& host,
+                                const int port) {
+  {
+    absl::WriterMutexLock lock{&cluster_map_lock_};
+    const auto cluster_it = cluster_map_.find(cluster_name);
+    if (cluster_it != cluster_map_.end()) {
+      cluster_it->second->touch();
+      return std::make_pair(true, absl::nullopt);
+    }
+    if (cluster_map_.size() >= max_sub_clusters_) {
+      ENVOY_LOG(debug, "cluster='{}' create failed due to max sub cluster limitation",
+                cluster_name);
+      return std::make_pair(false, absl::nullopt);
+    }
+    cluster_map_.emplace(cluster_name, std::make_shared<ClusterInfo>(cluster_name, *this));
+  }
+
+  // Inherit configuration from the parent DFP cluster.
+  envoy::config::cluster::v3::Cluster config = orig_cluster_config_;
+
+  // Overwrite the type.
+  config.set_name(cluster_name);
+  config.clear_cluster_type();
+  config.set_lb_policy(sub_cluster_lb_policy_);
+  config.set_type(
+      envoy::config::cluster::v3::Cluster_DiscoveryType::Cluster_DiscoveryType_STRICT_DNS);
+
+  // Set endpoint.
+  auto load_assignments = config.mutable_load_assignment();
+  load_assignments->set_cluster_name(cluster_name);
+  load_assignments->clear_endpoints();
+
+  auto socket_address = load_assignments->add_endpoints()
+                            ->add_lb_endpoints()
+                            ->mutable_endpoint()
+                            ->mutable_address()
+                            ->mutable_socket_address();
+  socket_address->set_address(host);
+  socket_address->set_port_value(port);
+
+  return std::make_pair(true, absl::make_optional(config));
+}
+
+Upstream::HostConstSharedPtr Cluster::chooseHost(absl::string_view host,
+                                                 Upstream::LoadBalancerContext* context) const {
+  uint16_t default_port = 80;
+  if (info_->transportSocketMatcher().resolve(nullptr).factory_.implementsSecureTransport()) {
+    default_port = 443;
+  }
+
+  const auto host_attributes = Http::Utility::parseAuthority(host);
+  auto dynamic_host = std::string(host_attributes.host_);
+  auto port = host_attributes.port_.value_or(default_port);
+
+  // cluster name is prefix + host + port
+  auto cluster_name = "DFPCluster:" + dynamic_host + ":" + std::to_string(port);
+
+  // try again to get the sub cluster.
+  auto cluster = cm_.getThreadLocalCluster(cluster_name);
+  if (cluster == nullptr) {
+    ENVOY_LOG(error, "cluster='{}' get thread local failed, too short ttl?", cluster_name);
```

Comment:
This could log spam, so let's not use an error log here. I'm not quite sure how this can reasonably happen so I would do a debug log and possibly an ENVOY_BUG if you are so inclined?

Reply:
I think it might be possible in such a case:
1. we touch the cluster with a short ttl, in the DFP filter, https://github.com/envoyproxy/envoy/blob/0497bd56aef3913ec09daafce3a047b5847fc0ff/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L303-L308
2. the cluster should be there, when first time chooseHost,
3. but it might be removed by the main thead, due to the short TTL,
4. and may fail to get the cluster, the next time chooseHost, when retry upstream request, after a bit long time.

this error log could only happens in a very low rate, and it worth an error level log, if the case is possible.

Reply:
I could see this only being possible if the TTL is unreasonably short (should there be some minimum in the config?). Either way, I don't think an error log is a code idea in the data plane path. Either switch to debug, or you could use a backoff warning log.

Reply:
it might be several seconds between two retry, may be up to the timeout configuration.
I'm not sure what could be a proper minimum value for the TTL.

Maybe, using a warning log could be good enough?